lib/moment_matching_da.py

```diff
@@ -0,0 +1,153 @@
+"""Domain adaptation via moment matching.
+"""
+import tensorflow as tf
+import tensorflow.keras as tkf
+
+class MomentMatchingClassifier(tkf.Model):
+  """Moments matching classifier for unsupervised domain adaptation.
+
+  Arguments:
+  classifier: classifier model.
+  target_domain: target domain name.
+  num_layers: number of resblock layers in the classifier.
+  """
+  def __init__(self, classifier, domains, target_domain,
+               num_layers, **args):
+    super(MomentMatchingClassifier, self).__init__(**args)
+    self.classifier = classifier
+    self.domains = domains
+    self.num_domains = len(self.domains)
+    self.target_domain = target_domain
+    if target_domain in domains:
+      self.target_domain_idx = int(self.domains.index(self.target_domain))
+    else:
+      raise ValueError("The target domain is not in the domains list.")
+
+    self.num_layers = num_layers
+    self._input_shape = self.classifier.input_shape[0][1:]
+    self._set_feature_extractor()
+
+
+  def _set_feature_extractor(self):
+    """Sets late feature extractor of the model."""
+    cls_inputs = [tkf.Input(name="%s_in_new" % x, shape=self._input_shape)
+                  for x in self.domains]
+    new_model_out = self.classifier(cls_inputs)
+    try:
+      block = (self.classifier.get_layer("feature_extractor")
+                              .get_layer("weighted_resblock"))
+      features_output_arr = []
+      for idx in range(self.num_domains):
+        out_idx = self.num_layers * (2 * self.num_domains - idx + 1) - 1
+        outputs = block.get_output_at(out_idx)
+        features_output_arr.append(outputs)
+    except ValueError:
+      #finding the last separate batch norm block
```

Comment:
This is another style guideline: make sure `#` is followed by a space in all comments.